compiler/rustc_errors/src/lib.rs

```diff
@@ -339,10 +345,12 @@ impl CodeSuggestion {
                     });
                     buf.push_str(&part.snippet);
                     let cur_hi = sm.lookup_char_pos(part.span.hi());
+                    trace!(?cur_hi);
                     if prev_hi.line == cur_lo.line {
                         // Account for the difference between the width of the current code and the
                         // snippet being suggested, so that the *later* suggestions are correctly
                         // aligned on the screen.
+                        trace!(?len);
                         acc += len as isize - (cur_hi.col.0 - cur_lo.col.0) as isize;
```

Comment:
The overflow happens here, because the lines of cur_hi and cur_lo are never checked against each other, so if lo starts in an earlier line but with a higher column, we get an overflow.

Now, we can properly fix this, and probably should, but the span that we have here is nonsensical to begin with. It goes from `bar(` inside the macro to the start of `true` at the *call site* of the macro.

Reply:
Ah! Yes, that's why we are getting these malformed spans. Yeah, macros screw everything up. Maybe we should have an `assert` (not an assert, a check) for well formed spans before we even try to emit or add a suggestion to the `DiagnosticBuilder`? That would make it so we never show suggestions like this one, while also not breaking the emitter logic. We should probably log that though.

Reply:
I _hate_ Spans sourced from macros 🥺

Reply:
We could add an actual `assert!` here. At least if anyone hits it we'll get a traceback.

Reply:
By the time we get here, the span doesn't look malformed anymore. The fault is either at the until method or at the caller to it. Maybe we should return an option instead of a span? Or a dummy span in case there are macros involved?

Reply:
We could tell that `lo > hi`, right? And in the Span wrangling methods we should check that both spans belong to the same Ctxt, and if not die in nightly and return only one of the original spans in beta/stable. Changing this to `Option` would be a royal pain in the neck.

Reply:
yea, debug asserting this would be best. We'll have to fix a bunch of diagnostics code that currently relies on it.

Reply:
I'm bumping into this overflow (again) in Clippy. Still not sure why it hasn't shown up in CI. But I narrowed it down to a suggestion that replaces an else block, so the block starts at a higher column than where it ends.

Reply:
@camsteffen yeah, didn't ping you on this PR when oli noticed this, but it is likely the same issue and this PR should help (but likely give... incorrect suggestions).

Reply:
I don't think the problem is only with invalid spans. Here is a minimal reproduction (triggers anonymous_parameters).

```rust
trait X {
    fn test(x: u32, (
    )) {}
}
```